internal/api/auth_hooks.go

```diff
@@ -0,0 +1,164 @@
+package api
+
+import (
+	"encoding/json"
+	"errors"
+	"net/url"
+
+	"fmt"
+	"github.com/gofrs/uuid"
+	"github.com/supabase/gotrue/internal/conf"
+	"github.com/supabase/gotrue/internal/storage"
+	"regexp"
+	"strings"
+)
+
+type HookType string
+
+const (
+	PostgresHook HookType = "pg-functions"
+	HTTPHook     HookType = "http"
+)
+
+const (
+	MFAHookRejection = "reject"
+	MFAHookContinue  = "continue"
+)
+
+type AuthHook struct {
+	ExtensibilityPointConfiguration conf.ExtensibilityPointConfiguration
+	payload                         []byte
+	hookType                        HookType
+	event                           string
+	db                              *storage.Connection
+}
+
+// Hook Events
+const (
+	MFAVerificationEvent = "auth.mfa_verfication"
+)
+
+type HookErrorResponse struct {
+	ErrorMessage string `json:"error_message"`
+	ErrorCode    string `json:"error_code"`
+	RetryAfter   bool   `json:"retry_after"`
+}
+
+func parseErrorResponse(response []byte) (*HookErrorResponse, error) {
+	var errResp HookErrorResponse
+	err := json.Unmarshal(response, &errResp)
+	if err != nil {
+		return nil, err
+	}
+	if errResp.ErrorMessage != "" {
+		return &errResp, nil
+	}
+	return nil, err
+}
+
+func parseMFAVerificationResponse(response []byte) (*MFAVerificationHookOutput, error) {
+	var MFAVerificationResponse MFAVerificationHookOutput
+	err := json.Unmarshal(response, &MFAVerificationResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MFAVerificationResponse, err
+}
+
+type MFAVerificationHookInput struct {
+	UserID   uuid.UUID `json:"user_id"`
+	FactorID uuid.UUID `json:"factor_id"`
+	Valid    bool      `json:"valid"`
+}
+type MFAVerificationHookOutput struct {
+	Decision string `json:"decision"`
+}
+
+// Functions for encoding and decoding payload
+func CreateMFAVerificationHookInput(user_id uuid.UUID, factor_id uuid.UUID, valid bool) ([]byte, error) {
+	payload := MFAVerificationHookInput{
+		UserID:   user_id,
+		FactorID: factor_id,
+		Valid:    valid,
+	}
+	data, err := json.Marshal(&payload)
+	if err != nil {
+		panic(err)
+	}
+	return data, nil
+}
```

Comment:
Let's avoid creating unnecessary functions to create structs. You should just be able to use `MFAVerificationInput` struct and fill it with things. 

Whether it gets serialized to JSON is implementation detail of the invoker of the hook; so that serialization / deserialization feels unnecessary to be mandatory here.

Reply:
Ah good point, thanks!